File: scala/scalaz/arrow/move_knight/src/main/scala/MoveKnight.scala
package sample

import scalaz._
import Scalaz._

object MoveKnight extends App {

	type KnightPos = Tuple2[Int, Int]

	val inRange = (p: Int) => 1 to 8 contains p

	val moveKnight = (p: KnightPos) => List(
		(p._1 + 2, p._2 - 1), (p._1 + 2, p._2 + 1),
		(p._1 - 2, p._2 - 1), (p._1 - 2, p._2 + 1),
		(p._1 + 1, p._2 - 2), (p._1 + 1, p._2 + 2),
		(p._1 - 1, p._2 - 2), (p._1 - 1, p._2 + 2)
	).filter { case (x, y) => inRange(x) && inRange(y) }

	val in3 = Kleisli(moveKnight) >==> moveKnight >==> moveKnight

	in3 (6, 2) |> println

	val canReachIn3 = (end: KnightPos) => in3.run >>> { xs => xs.contains(end) }

	(6, 2) |> canReachIn3 (6, 1) |> println
	(6, 2) |> canReachIn3 (7, 3) |> println
}

File: Haskell/arrow/move_knight/move_knight.hs
import Control.Arrow

type KnightPos = (Int, Int)

moveKnight :: KnightPos -> [KnightPos]
moveKnight (c, r) = filter onBoard
	[
		(c + 2, r - 1), (c + 2, r + 1),
		(c - 2, r - 1), (c - 2, r + 1),
		(c + 1, r - 2), (c + 1, r + 2),
		(c - 1, r - 2), (c - 1, r + 2)
	]
	where onBoard (c', r') = c' `elem` [1..8] && r' `elem` [1..8]

in3 = Kleisli moveKnight >>> Kleisli moveKnight >>> Kleisli moveKnight

canReachIn3 end = runKleisli in3 >>> elem end

main = do
	print $ runKleisli in3 $ (6, 2)

	print $ canReachIn3 (6, 1) $ (6, 2)
	print $ canReachIn3 (7, 3) $ (6, 2)
